### PR TITLE
Added argument forwarding if_set, if_unset, value_or and and_then

### PR DESCRIPTION
**Pull request by gotbordom**

### CODE REVIEW

File: safet/optional.hpp [hunk 1]
```diff
@@ -161,182 +162,189 @@ class optional {
...
+   template <typename Functor, typename... AdditionalArgs>
+    decltype(auto) if_set(Functor&& f, AdditionalArgs&&... additional_args) const&
     {
-        static_assert(impl::invocable<Functor&&, const T&>, "if_set functor on const optional must be invocable with const T&");
+        static_assert(impl::invocable<Functor&&, const T&, const AdditionalArgs&&...>, "if_set functor on const optional must be invocable with const T&");
 
-        if constexpr (impl::invocable_and_returns_something<Functor&&, const T&>) {
-            return [&]() -> optional<std::invoke_result_t<Functor&&, const T&>> {
+        if constexpr (impl::invocable_and_returns_something<Functor&&, const T&, const AdditionalArgs&&...>) {
+            return [&]() -> optional<std::invoke_result_t<Functor&&, const T&, const AdditionalArgs&&...>> {
```
Comment (kevin-logan) [inline safet/optional.hpp]: the consts here should not be applied to AdditionalArgs (only to the optional's value), constifying the forwarding references will break forwarding, a unit test where one of the additional arguments is a move-only type, like an std::unique_ptr, could be used to ensure forwarding works in more cases

File: tests/optional.cpp [hunk 2]
```diff
@@ -334,6 +383,20 @@ TEST_CASE("optional::if_unset()", "[optional]")
...
+        optional o { 1 };
+        auto test_argument = 2;
+
+        // functor must be called if optional is disengaged
+        o = std::nullopt;
+        auto result = o.if_unset([](auto& test_argument) { return 2 + test_argument; }, test_argument);
+        REQUIRE(result == 4);
+
+        // result must be set now, and should be 4
+        result.if_set([](auto val) { REQUIRE(val == 4); }).if_unset([]() { FAIL("if_unset must not be called on an engaged optional"); });
```
Comment (kevin-logan) [inline tests/optional.cpp]: both of these conditions are already implicitly checked with the `REQUIRE(result == 4)` above

File: tests/optional.cpp [hunk 3]
```diff
@@ -392,6 +475,43 @@ TEST_CASE("optional::and_then()", "[optional]")
...
+
+        // empty optional if the functor returns nullopt
+        s = "not a number";
+        i = s.and_then([&](auto value, auto& test_argument) { return sum_string_and_digit(value, test_argument); }, test_argument);
+        REQUIRE(i.empty());
+
+        // "5" will succeed in string_to_digit, so the conversion must succeed
+        s = "5";
+        i = s.and_then([&](auto value, auto& test_argument) { return sum_string_and_digit(value, test_argument); }, test_argument);
+        i.if_set([](int value) { REQUIRE(value == 7); }).if_unset([]() { FAIL("and_then must return an engaged optional when the optional is engaged and the functor succeeds"); });
```
Comment (kevin-logan) [inline tests/optional.cpp]: prefer just using the syntax `REQUIRE(i == 7)` here, it's a lot more succinct and ultimately checks the same edge cases. I know you were just copying this from above but unfortunately some of these tests were written before the == operator

File: tests/optional.cpp [hunk 4]
```diff
@@ -392,6 +475,43 @@ TEST_CASE("optional::and_then()", "[optional]")
...
+
+        optional move_only { std::make_unique<std::string>("3") };
+
+        // move semantics work
+        i = std::move(move_only).and_then([&](auto value_ptr, auto& test_argument) { return sum_string_and_digit(*value_ptr, test_argument); }, test_argument);
+        i.if_set([](int value) { REQUIRE(value == 5); }).if_unset([]() { FAIL("and_then must return an engaged optional when the optional is engaged and the functor succeeds"); });
```
Comment (kevin-logan) [inline tests/optional.cpp]: same, prefer `REQUIRE(i == 5)`

**Review by kevin-logan — CHANGES_REQUESTED** [review]

File: safet/optional.hpp (hunk 1 above)
Comment (gotbordom) [reply safet/optional.hpp]: This const should be the case for all the changes I made correct? I didn't realize this until you mentioned it so I did some googling and in all the const function cases I passed const AdditionalArgs so I think all of those are broken because I guess argument forwarding relies on move semantics which can't be moved if we have const references?

Comment (kevin-logan) [reply safet/optional.hpp]: right, there are a few places right here where you're adding "const" to "AdditionalArgs", you don't want that (because it means AdditionalArgs can't be moved when appropriate). I didn't see it in any other functions, though.

Comment (gotbordom) [reply safet/optional.hpp]: Looking at the code I guess I didn't but I really thought I did. So I'll fix this and ad a unit tests

File: tests/optional.cpp (hunk 4 above)
Comment (gotbordom) [reply tests/optional.cpp]: For these I noticed there are other places in the code that I had not changed that I could also update to match this. Do you still want those cases to throw errors and just only have the "section(argument forwarding)" to use this format? Or should all unit tests just use this format?

Comment (kevin-logan) [reply tests/optional.cpp]: you're free to change them anywhere you see that pattern, but I just meant for the new code you've introduced

File: tests/optional.cpp [hunk 5]
```diff
@@ -316,6 +316,64 @@ TEST_CASE("optional::if_set()", "[optional]")
...
+        }, test_argument);
+
+        // check the const method of if_set works, also check that we can pass move only parameters
+        const optional const_o { 4 };
+        const auto& new_const_o = const_o.if_set([](auto& value, auto test_pointer)
+        {
+            REQUIRE(value == 4);
+            REQUIRE(*test_pointer == 10);
+            return value;
+        }, std::move(std::make_unique<int>(10)));
```
Comment (kevin-logan) [inline tests/optional.cpp]: this move here is unnecessary, as std::make_unique's return value is implicitly an r-value

**Review by kevin-logan — APPROVED** [review]
